src/com/amazon/ion/benchmark/WriteRandomIonValues.java

```diff
@@ -218,122 +247,176 @@ private static Integer randomLocalOffset(Random random) {
-                    throw new IllegalStateException("Found non-timestamp");
+//            Timestamp.Precision[] precisions = Timestamp.Precision.values();
+            if (timestampTemplate != "random"){
+                while (size > 0) {
+                    IonReaderBuilder readerBuilder = IonReaderBuilder.standard();
+                    IonReader reader = readerBuilder.build(timestampTemplate);
+                    reader.next();
+                    reader.stepIn();
+                    reader.next();
+                    int count = 0;
+                    while (reader.isNullValue() == false){
+                        count += 1;
+                        Timestamp value = reader.timestampValue();
+                        Timestamp.Precision precision = value.getPrecision();
+                        Timestamp timestamp;
+                        Random random = new Random();
+
+                        switch (precision) {
+                            case YEAR:
+                                timestamp = Timestamp.forYear(random.nextInt(9998) + 1);
+                                break;
+                            case MONTH:
+                                timestamp = Timestamp.forMonth(random.nextInt(9998) + 1, random.nextInt(12) + 1);
+                                break;
+                            case DAY:
+                                timestamp = Timestamp.forDay(random.nextInt(9998) + 1, random.nextInt(12) + 1,
+                                        random.nextInt(28) + 1 // Use max 28 for simplicity. Not including up to 31 is not going to
+                                        // affect the measurement.
+                                );
+                                break;
+                            case MINUTE:
+                                timestamp = Timestamp.forMinute(random.nextInt(9998) + 1, random.nextInt(12) + 1,
+                                        random.nextInt(28) + 1, // Use max 28 for simplicity. Not including up to 31 is not going to
+                                        // affect the measurement.
+                                        random.nextInt(24), random.nextInt(60), randomLocalOffset(random));
+                                break;
+                            case SECOND:
+                                //added
+                                BigDecimal secondValue = value.getDecimalSecond();
+                                int scale = secondValue.scale();
+                                timestamp = Timestamp.forSecond(random.nextInt(9998) + 1, random.nextInt(12) + 1,
+                                        random.nextInt(28) + 1, // Use max 28 for simplicity. Not including up to 31 is not going to
+                                        // affect the measurement.
+                                        random.nextInt(24), random.nextInt(60), randomSecondWithFraction(random,scale), randomLocalOffset(random));
```

Comment:
Have you considered how you will generate fractional seconds and local offset from the template as well?